middleware: Add client version to log and other logging improvements.

orientor opened the pull request:
The first commit uses get_client_name in LogRequests's process_request and removes its usage in process_client.
The second commit adds request.client_version to LogRequest's process_request.
The third commit changes write_log_line so that it now shows client version if available. 
The fourth commit uses client id instead of client name for logging.

Fixes #14067 .

timabbott commented on zerver/middleware.py:
```diff
@@ -250,6 +252,13 @@ def process_request(self, request: HttpRequest) -> None:
+            request.client_name = get_client_name(request, is_browser_view=True)
+            request.client_version = None
+            if request.client_name.startswith("Zulip"):
+                request.client_version = parse_user_agent(request.META["HTTP_USER_AGENT"])["version"]
```
This should be an independent function taking the `request` as its argument; defining a function inside another function means we do work to create the function every time `process_request` is called.

orientor replied:
Made it an independent method of class LogRequests.

timabbott commented on zerver/decorator.py:
```diff
@@ -159,7 +159,10 @@ def process_client(request: HttpRequest, user_profile: UserProfile,
+        if hasattr(request, "client_name"):
+            client_name = request.client_name
+        else:
+            client_name = get_client_name(request, is_browser_view=is_browser_view)
```
When do you expect this `else` case to happen?

orientor replied:
I am not sure why this happens but if I don't put the else part then during the tests I get:
```
Running zerver.tests.test_decorators.DecoratorLoggingTestCase.test_authenticated_rest_api_view_with_non_webhook_view

======================================================================
FAIL: test_authenticated_rest_api_view_with_non_webhook_view (zerver.tests.test_decorators.DecoratorLoggingTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.5/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/usr/lib/python3.5/unittest/case.py", line 600, in run
    testMethod()
  File "/home/circleci/zulip/zerver/tests/test_decorators.py", line 575, in test_authenticated_rest_api_view_with_non_webhook_view
    non_webhook_view_raises_exception(request)
  File "/usr/lib/python3.5/unittest/case.py", line 212, in __exit__
    expected_regex.pattern, str(exc_value)))
  File "/usr/lib/python3.5/unittest/case.py", line 134, in _raiseFailure
    raise self.test_case.failureException(msg)
AssertionError: "raised by a non-webhook view" does not match "'HostRequestMock' object has no attribute 'client_name'"

```
Most probably this test directly passes a request variable to process_client without calling the middleware. Will look into it.

timabbott commented on zerver/middleware.py:
```diff
@@ -250,6 +252,10 @@ def process_request(self, request: HttpRequest) -> None:
             return
 
+        def process_user_agent() -> None:
+            request.client_name = get_client_name(request, is_browser_view=True)
```
It is not correct to hardcode `is_browser_view=True`.  

I think what we need to do to resolve this is to refactor the `is_browser_view` logic to be done in `process_client`, rather than `get_client_name`.

timabbott replied:
OK, I figured out how to do this in a reasonable way; I've merged a series of prep commits that take care of this problem by removing that argument.  See the series ending with ccf63ac66bdaa221c8859f711f8e7d53a834fc58.

orientor replied:
Thanks for cleaning up get_client_name. I have accordingly rebased my commits and now the codeflow seems smoother.

timabbott commented:
Thanks for working on this @orientor!  I posted several comments and spent a couple hours cleaning up the `get_client_name` interface (which had some subtle history underlying it) in a way should make it possible to do this in a reasonable way.  Can you rebase and update your commits for my comments above?

orientor commented:
@timabbott added another commit so that now client version is displayed in logging. Also changed the tests so that we don't need the else part(which would never run in real life) and changed code according to review. Kindly check once again. I think it should be mergeable after the tests are back to normal.

mateuszmandera commented on zerver/middleware.py:
```diff
@@ -306,6 +309,10 @@ def process_response(self, request: HttpRequest,
+        try:
+            client_version = request.client_version
+        except Exception:
+            client_version = None
```
When would we expect an `Exception` to happen here?

orientor replied:
According to me in any case as the corresponding process_request will always run before the process_response, so we should always have the client_version attribute. But in the circleci
and even on my development server I am getting this error:

```
2020-03-15 18:15:12.563 ERR  [django.request] Internal Server Error: /json/events
Traceback (most recent call last):
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/utils/deprecation.py", line 96, in __call__
    response = self.process_response(request, response)
  File "/home/circleci/zulip/zerver/middleware.py", line 312, in process_response
    client_version = request.client_version
AttributeError: 'WSGIRequest' object has no attribute 'client_version'
```

I couldn't find why the error occured. It would be great if you can help.

mateuszmandera replied:
Ah, I know what the problem is. In our magical `AsyncDjangoHandler` in `zerver/tornado/handlers.py` we do some fiddling with request objects and need to copy over these special attributes. This block:
```
        # Add to this new HttpRequest logging data from the processing of
        # the original request; we will need these for logging.
        #
        # TODO: Design a cleaner way to manage these attributes,
        # perhaps via creating a ZulipHttpRequest class that contains
        # these attributes with a copy method.
        request._log_data = old_request._log_data
        if hasattr(request, "_rate_limit"):
            request._rate_limit = old_request._rate_limit
        if hasattr(request, "_requestor_for_logs"):
            request._requestor_for_logs = old_request._requestor_for_logs
        request.user = old_request.user
        request.client = old_request.client
```

we should copy both the `client_name` and `client_version` from old_request to request in the same way. And then this `except Exception` should be able to be removed.

orientor replied:
Thanks for debugging. Will get to it.

orientor replied:
@mateuszmandera fixed it. Removed the try catch in the client_name part too.

mateuszmandera replied:
Thanks, looks good!

mateuszmandera commented:
I looked through the commits, and seems good to me! One alternative idea I have here, that maybe @timabbott can first say if it would be an improvement, is that perhaps instead of doing the `request.client_name/version` setting in logging middleware, we should make a new `UserAgentMiddle`, that would run before logging middleware, and do these things? The reasoning is that these changes make the logging middleware do other stuff than simply logging, stuff that one probably wouldn't expect from the name and assumed responsibilities of this class.

orientor commented:
@mateuszmandera nice idea. It would make the code more coherent and easier to document.

timabbott commented:
@orientor can you rebase?

orientor commented:
@timabbott rebased the PR.

timabbott commented:
The last commit is wrong, and I've removed it.  If we're going to change that interface, we have to change the queue processor as well.  @mateuszmandera I'd prefer that you do the work for replacing `client` with `client_id` in the interface for UserActivityWorker; it's really sensitive production-focused work.

I pushed back the other changes.  One mystery that needs to be resolved before we can merge this is that we start logging as "Mozilla" rather than "website" in `run-dev.py` logs after the first commit is added, and I don't see why that would be the case here but not in master.

mateuszmandera commented:
> I pushed back the other changes. One mystery that needs to be resolved before we can merge this is that we start logging as "Mozilla" rather than "website" in `run-dev.py` logs after the first commit is added, and I don't see why that would be the case here but not in master.

@timabbott  Added the following diff to the first commit and rebased:
```
diff --git a/zerver/middleware.py b/zerver/middleware.py
index 18b15c3cb2..da74323adc 100644
--- a/zerver/middleware.py
+++ b/zerver/middleware.py
@@ -299,7 +299,10 @@ class LogRequests(MiddlewareMixin):
             else:
                 requestor_for_logs = "unauth@%s" % (get_subdomain(request) or 'root',)

-        client = request.client_name
+        try:
+            client = request.client.name
+        except Exception:
+            client = request.client_name

         if response.streaming:
             content_iter = response.streaming_content
```

The reason for the problem was that we do the conversion frmo "Mozilla" etc. in `process_client`:
```
    # We could check for a browser's name being "Mozilla", but
    # e.g. Opera and MobileSafari don't set that, and it seems
    # more robust to just key off whether it was a browser view
    if is_browser_view and not client_name.startswith("Zulip"):
        # Avoid changing the client string for browsers, but let
        # the Zulip desktop apps be themselves.
        client_name = "website"
```

And we need the `is_browser_view` informatoin, so I can't see a good way to do this differently.
My changes make us have `website` again, except in views like `/login/` and `/devlogin/`, because those codepaths don't call `process_client`, so `request.client_name` will be used the logging middleware and thus print out `Mozilla`.

orientor commented:
@mateuszmandera thanks for editing.Was a little busy during the previous days due to uni restarting.

mateuszmandera commented:
@orientor No problem! :slightly_smiling_face: 

@timabbott Added the client_id change and a fix for rather unfortunate typos in the test for that queue worker.

timabbott commented:
I merged a couple commits from this branch.  I don't understand the reasoning for why the remaining changes are correct, which maybe should go into a comment -- e.g. why is it safe to assume `client_version` is always present, for example?

timabbott commented:
Closing in favor of #17800.
